wire/merkle_tree.go

```diff
@@ -0,0 +1,310 @@
+package wire
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/tokenized/pkg/bitcoin"
+)
+
+// MerkleTree is an efficient structure for calculating a merkle root hash.
+type MerkleTree struct {
+	layers []*merkleNodeLayer // First layer, index zero, is the lowest level of the tree.
+	prune  bool
+	count  int
+
+	merkleProofs []*MerkleProof
+}
+
+func NewMerkleTree(prune bool) *MerkleTree {
+	return &MerkleTree{
+		prune: prune,
+	}
+}
+
+// AddMerkleProof adds a merkle proof to be calculated with the tree.
+func (t *MerkleTree) AddMerkleProof(txid bitcoin.Hash32) {
+	t.merkleProofs = append(t.merkleProofs, NewMerkleProof(txid))
+}
+
+func (t MerkleTree) Print() {
+	indent := ""
+	for i, layer := range t.layers {
+		fmt.Printf("%sLayer %d\n", indent, i+1)
+		indent += "  "
+		hashIndex := layer.count - len(layer.hashes) + 1
+		for _, hash := range layer.hashes {
+			fmt.Printf("%s%d : %s\n", indent, hashIndex, hash.String())
+			hashIndex++
+		}
+	}
+}
+
+// AddHash adds a new hash to the merkle tree.
+func (t *MerkleTree) AddHash(hash bitcoin.Hash32) {
+
+	// Check merkle proofs for this hash.
+	for _, mp := range t.merkleProofs {
+		if mp.Index == -1 && mp.TxID != nil && mp.TxID.Equal(&hash) {
+			mp.Index = t.count
+			break
+		}
+	}
+
+	if len(t.layers) == 0 {
+		// First hash in tree
+		t.layers = []*merkleNodeLayer{newMerkleNodeLayer(hash)}
+		t.count = 1
+		return
+	}
+
+	next := hash
+	t.count++
+
+	// Calculate a new hash up the tree
+	s := sha256.New()
+	for _, layer := range t.layers {
+		// Append to this row
+		layer.addHash(next)
+
+		l := layer.len()
+		if l%2 != 0 {
+			return // Above layers do not need to be updated.
+		}
+
+		left := layer.nextLastHash()
+		right := next
+
+		// Even number of hashes in layer. Hash last 2 hashes together to add to layer above.
+		s.Write(left[:])
+		s.Write(right[:])
+		b := s.Sum(nil)
+		s.Reset()
+		next = bitcoin.Hash32(sha256.Sum256(b)) // Sum again for double SHA256
+		if t.prune {
+			layer.clear() // Clear out hashes that aren't needed anymore
+		}
+
+		// Check active merkle proofs for next element in path
+		for _, mp := range t.merkleProofs {
+			if mp.Index == -1 {
+				continue // txid not found yet
+			}
+
+			if mp.root.Equal(&left) {
+				mp.AddHash(right, next)
+			} else if mp.root.Equal(&right) {
+				mp.AddHash(left, next)
+			}
+		}
+	}
+
+	// Append new layer
+	t.layers = append(t.layers, newMerkleNodeLayer(next))
+}
+
+func (t *MerkleTree) RootHash() bitcoin.Hash32 {
```

Comment:
Please add comment.

Reply:
I am not sure why it wouldn't be obvious that it returns the root hash of the merkle tree, but I will add a comment.